flax/linen/normalization.py

```diff
@@ -849,3 +849,169 @@ def _spectral_normalize(self, path, vs, layer_instance_name, update_stats):
+      ``layer_instance`` variables whose key path (delimited by '/') has a
+      match with ``variable_filter``. For example, ``variable_filter={'kernel'}``
+      will only apply l2-normalization to variables whose key path contains
+      'kernel'. By default, ``variable_filter={'kernel'}``.
+  """
+
+  layer_instance: Module
+  epsilon: float = 1e-12
+  dtype: Optional[Dtype] = None
+  param_dtype: Dtype = jnp.float32
+  use_scale: bool = True
+  scale_init: Callable[[PRNGKey, Shape, Dtype], Array] = initializers.ones
+  feature_axes: Optional[Axes] = -1
+  variable_filter: Optional[Iterable] = field(default_factory=lambda: {'kernel'})
+
+  @compact
+  def __call__(self, *args, **kwargs):
+    """Compute the l2-norm of the weights in ``self.layer_instance``
+    and normalize the weights using this value before computing the
+    ``__call__`` output.
+
+    Args:
+      *args: positional arguments to be passed into the call method of the
+        underlying layer instance in ``self.layer_instance``.
+      **kwargs: keyword arguments to be passed into the call method of the
+        underlying layer instance in ``self.layer_instance``.
+
+    Returns:
+      Output of the layer using l2-normalized weights.
+    """
+
+    def layer_forward(layer_instance):
+      return layer_instance(*args, **kwargs)
+
+    return map_variables(
+        layer_forward,
+        trans_in_fn=lambda vs: jax.tree_util.tree_map_with_path(
+          self._l2_normalize,
+          vs,
+        ),
+        init=self.is_initializing(),
+    )(self.layer_instance)
+
+  def _l2_normalize(self, path, vs):
+    """Compute the l2-norm and normalize the variables ``vs`` using this
+    value. This is intended to be a helper function used in this Module's
+    ``__call__`` method in conjunction with ``nn.transforms.map_variables``
+    and ``jax.tree_util.tree_map_with_path``.
+
+    Args:
+      path: dict key path, used for naming the ``scale`` variable
+      vs: variables to be l2-normalized
+    """
+    value = jnp.asarray(vs)
+    str_path = self.layer_instance.name + '/' + '/'.join((dict_key.key for dict_key in path[1:])) + '/scale'
+    if self.variable_filter:
+      for variable_name in self.variable_filter:
+        if variable_name in str_path:
+          break
+      else:
+        return value
+
+    if self.feature_axes is None:
+      feature_axes = ()
+      reduction_axes = tuple(i for i in range(value.ndim))
+    else:
+      feature_axes = _canonicalize_axes(value.ndim, self.feature_axes)
+      reduction_axes = tuple(i for i in range(value.ndim) if i not in feature_axes)
+
+    feature_shape = [1] * value.ndim
+    reduced_feature_shape = []
+    for ax in feature_axes:
+      feature_shape[ax] = value.shape[ax]
+      reduced_feature_shape.append(value.shape[ax])
+
+    value_bar = jnp.where(value != 0,
+                          _l2_normalize(value, axis=reduction_axes, eps=self.epsilon),
+                          value)
```

Comment:
Can you explain what is going on here? Seems to be keeping the zeros but now the norm is not `1`.

Reply:
I removed it